pkg/services/ngalert/notifier/images.go

```diff
@@ -6,151 +6,73 @@ import (
 	"io"
 	"os"
 	"path/filepath"
-	"strings"
 
 	alertingImages "github.com/grafana/alerting/images"
-	alertingModels "github.com/grafana/alerting/models"
-	alertingNotify "github.com/grafana/alerting/notify"
 
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/services/ngalert/models"
 	"github.com/grafana/grafana/pkg/services/ngalert/store"
 )
 
-type imageProvider struct {
+type tokenStore struct {
 	store  store.ImageStore
 	logger log.Logger
 }
 
+var _ alertingImages.TokenStore = (*tokenStore)(nil)
+
 func newImageProvider(store store.ImageStore, logger log.Logger) alertingImages.Provider {
-	return &imageProvider{
+	return alertingImages.NewTokenProvider(&tokenStore{
 		store:  store,
 		logger: logger,
-	}
+	}, newLogWrapper(logger))
 }
 
-func (i imageProvider) GetImage(ctx context.Context, uri string) (*alertingImages.Image, error) {
-	image, err := i.getImageFromURI(ctx, uri)
+func (t tokenStore) GetImage(ctx context.Context, token string) (*alertingImages.Image, error) {
+	image, err := t.store.GetImage(ctx, token)
 	if err != nil {
 		if errors.Is(err, models.ErrImageNotFound) {
-			i.logger.Info("Image not found in database", "uri", uri)
 			return nil, alertingImages.ErrImageNotFound
 		}
 		return nil, err
 	}
 
 	return &alertingImages.Image{
-		Token:     image.Token,
-		Path:      image.Path,
-		URL:       image.URL,
-		CreatedAt: image.CreatedAt,
+		URL: image.URL,
+		RawData: func(_ context.Context) (alertingImages.ImageContent, error) {
+			if image.Path == "" {
+				return alertingImages.ImageContent{}, models.ErrImageDataUnavailable
+			}
+			b, err := readImage(image.Path, t.logger)
+			if err != nil {
+				return alertingImages.ImageContent{}, err
+			}
+			return alertingImages.ImageContent{
+				Name:    filepath.Base(image.Path),
+				Content: b,
+			}, nil
+		},
```

Comment:
This method can be accessed concurrently, for example, when receiver has multiple integrations. WDYT about implementing a pattern when file can only be read once?